File: modules/rounds-native/android/src/main/java/app/rounds/nativemodule/RoundsShared.kt
package app.rounds.nativemodule

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

/**
 * The shared store every out-of-app surface writes into.
 *
 * A notification action, a Glance widget and a Quick Settings tile all run
 * outside the React Native process — none of them can reach the JS offline
 * queue while the app is dead. So they append here, and the app drains it on
 * next foreground.
 *
 * The critical detail: **the surface mints the UUID**, not the app. That is
 * what makes the whole thing idempotent. Tapping "Same again" three times on a
 * notification with no network produces three rows with three ids; a drain that
 * runs twice produces no duplicates at all.
 */
object RoundsShared {
    private const val PREFS = "rounds_shared"
    private const val KEY_PENDING = "rounds.pending"
    private const val KEY_WIDGET = "rounds.widget"
    private const val KEY_HUD = "rounds.hud"
    private const val KEY_TILE_DRINK = "rounds.tile.drinkId"
    private const val KEY_TILE_LABEL = "rounds.tile.label"

    /** A phone offline for a week must not hand the app ten thousand rows. */
    private const val MAX_PENDING = 500

    private fun prefs(context: Context): SharedPreferences =
        context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    data class PendingLog(val id: String, val drinkId: String, val at: Long, val source: String) {
        fun toJson(): JSONObject = JSONObject()
            .put("id", id).put("drinkId", drinkId).put("at", at).put("source", source)

        companion object {
            fun create(drinkId: String, source: String) =
                PendingLog(UUID.randomUUID().toString(), drinkId, System.currentTimeMillis(), source)

            fun fromJson(o: JSONObject) = PendingLog(
                o.optString("id"), o.optString("drinkId"),
                o.optLong("at"), o.optString("source")
            )
        }
    }

    @Synchronized
    fun appendPending(context: Context, log: PendingLog) {
        val all = pending(context).toMutableList()
        all.add(log)
        val bounded = if (all.size > MAX_PENDING) all.takeLast(MAX_PENDING) else all
        val array = JSONArray()
        bounded.forEach { array.put(it.toJson()) }
        prefs(context).edit().putString(KEY_PENDING, array.toString()).apply()
    }

    fun pending(context: Context): List<PendingLog> {
        val raw = prefs(context).getString(KEY_PENDING, null) ?: return emptyList()
        return runCatching {
            val array = JSONArray(raw)
            (0 until array.length()).map { PendingLog.fromJson(array.getJSONObject(it)) }
        }.getOrDefault(emptyList())
    }

    /** Called only after the app has handed the rows to the queue. */
    fun clearPending(context: Context) {
        prefs(context).edit().remove(KEY_PENDING).apply()
    }

    fun writeWidget(context: Context, json: String) {
        prefs(context).edit().putString(KEY_WIDGET, json).apply()
    }
    fun widget(context: Context): String? = prefs(context).getString(KEY_WIDGET, null)

    fun writeHud(context: Context, json: String) {
        prefs(context).edit().putString(KEY_HUD, json).apply()
    }
    fun hud(context: Context): String? = prefs(context).getString(KEY_HUD, null)

    fun writeTile(context: Context, drinkId: String, label: String) {
        prefs(context).edit()
            .putString(KEY_TILE_DRINK, drinkId)
            .putString(KEY_TILE_LABEL, label)
            .apply()
    }
    fun tileDrinkId(context: Context): String =
        prefs(context).getString(KEY_TILE_DRINK, "beer-pint") ?: "beer-pint"
    fun tileLabel(context: Context): String =
        prefs(context).getString(KEY_TILE_LABEL, "Log a drink") ?: "Log a drink"
}

/**
 * The HUD state. Mirrors `RoundsHudState` in Swift and `HudState` in TypeScript.
 *
 * Note what is NOT here: the ‰ estimate. An ongoing notification sits on a lock
 * screen anyone can see, and that is not a place for a number that invites the
 * one interpretation the product must never invite.
 */
data class RoundsHud(
    val sessionId: String,
    val venue: String?,
    val startedAt: Long,
    val drinks: Int,
    val paceState: String,
    val paceWord: String,
    val lastDrinkId: String?,
    val lastDrinkName: String?,
) {
    val accentColor: Int
        get() = when (paceState) {
            "steady" -> 0xFF30D158.toInt()
            "quick" -> 0xFFFF9F0A.toInt()
            "slow_down" -> 0xFFFF453A.toInt()
            else -> 0xFF7CB3FF.toInt()
        }

    val elapsedLabel: String
        get() {
            val mins = ((System.currentTimeMillis() - startedAt) / 60000).coerceAtLeast(0)
            return if (mins >= 60) "${mins / 60}h${(mins % 60).toString().padStart(2, '0')}" else "${mins}m"
        }

    companion object {
        fun parse(json: String): RoundsHud? = runCatching {
            val o = JSONObject(json)
            RoundsHud(
                sessionId = o.getString("sessionId"),
                venue = o.optString("venue").takeIf { it.isNotEmpty() && it != "null" },
                startedAt = o.optLong("startedAt"),
                drinks = o.optInt("drinks"),
                paceState = o.optString("paceState", "easy"),
                paceWord = o.optString("paceWord", "EASY"),
                lastDrinkId = o.optString("lastDrinkId").takeIf { it.isNotEmpty() && it != "null" },
                lastDrinkName = o.optString("lastDrinkName").takeIf { it.isNotEmpty() && it != "null" },
            )
        }.getOrNull()
    }
}

File: modules/rounds-native/ios/RoundsShared.swift
import Foundation

/// The shared container every out-of-app surface writes into.
///
/// A Live Activity button, a widget, a Control Center control and the watch all
/// run in separate processes from the app, and none of them can reach the JS
/// offline queue while the app is suspended. So they append here — an App Group
/// UserDefaults — and the app drains it on next foreground.
///
/// The critical detail: **the surface mints the UUID**, not the app. That is
/// what makes the whole thing idempotent. Tapping "Same again" three times on a
/// Lock Screen with no network produces three rows with three ids; a drain that
/// runs twice produces no duplicates at all.
public enum RoundsShared {
  public static let appGroup = "group.app.rounds.client"

  private static var defaults: UserDefaults? { UserDefaults(suiteName: appGroup) }

  private static let pendingKey = "rounds.pending"
  private static let widgetKey  = "rounds.widget"
  private static let hudKey     = "rounds.hud"

  public struct PendingLog: Codable {
    public let id: String
    public let drinkId: String
    public let at: Double        // epoch ms, to match the JS layer exactly
    public let source: String

    public init(drinkId: String, source: String) {
      self.id = UUID().uuidString.lowercased()
      self.drinkId = drinkId
      self.at = Date().timeIntervalSince1970 * 1000
      self.source = source
    }
  }

  /// Append a log from any surface. Never blocks, never reaches the network.
  public static func appendPending(_ log: PendingLog) {
    guard let defaults else { return }
    var all = pending()
    all.append(log)
    // Bounded: a phone that has been offline for a week should not be able to
    // hand the app ten thousand rows on a single foreground.
    if all.count > 500 { all = Array(all.suffix(500)) }
    if let data = try? JSONEncoder().encode(all) {
      defaults.set(data, forKey: pendingKey)
    }
  }

  public static func pending() -> [PendingLog] {
    guard let defaults, let data = defaults.data(forKey: pendingKey) else { return [] }
    return (try? JSONDecoder().decode([PendingLog].self, from: data)) ?? []
  }

  /// Called only after the app has handed the rows to the queue.
  public static func clearPending() {
    defaults?.removeObject(forKey: pendingKey)
  }

  public static func writeWidgetPayload(_ json: String) {
    defaults?.set(json, forKey: widgetKey)
  }

  public static func widgetPayload() -> String? {
    defaults?.string(forKey: widgetKey)
  }

  public static func writeHud(_ json: String) { defaults?.set(json, forKey: hudKey) }
  public static func hud() -> String? { defaults?.string(forKey: hudKey) }
}

/// The HUD's state, shared between the app, the Live Activity and the watch.
///
/// Note what is NOT in here: the ‰ estimate. A Live Activity sits on a Lock
/// Screen anyone can see over your shoulder, and that is not a place for a
/// number that invites the one interpretation the product must never invite.
/// The surfaces carry the state word and the count.
public struct RoundsHudState: Codable, Hashable {
  public let sessionId: String
  public let venue: String?
  public let startedAt: Double
  public let drinks: Int
  public let paceState: String   // easy | steady | quick | slow_down
  public let paceWord: String    // EASY | STEADY | QUICK | SLOW DOWN
  public let lastDrinkId: String?
  public let lastDrinkName: String?
  public let spendMinor: Int
  public let currency: String

  public var elapsedMinutes: Int {
    max(0, Int((Date().timeIntervalSince1970 * 1000 - startedAt) / 60000))
  }

  public var accentHex: String {
    switch paceState {
    case "steady":    return "#30D158"
    case "quick":     return "#FF9F0A"
    case "slow_down": return "#FF453A"
    default:          return "#7CB3FF"
    }
  }
}
